models/migrations/v58.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2017 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"fmt"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+
+	"github.com/go-sql-driver/mysql"
+	"github.com/go-xorm/xorm"
+)
+
+func mysqlColumnsToUTF8MB4(x *xorm.Engine) (err error) {
+	if !setting.UseMySQL {
+		log.Info("Nothing to do")
+		return nil
+	}
+
+	const maxvc = 191
+	migrationSuccess := true
+
+	tables, err := x.DBMetas()
+	if err != nil {
+		return fmt.Errorf("cannot get tables: %v", err)
+	}
+	for _, table := range tables {
+		readyForConversion := true
+		for ix := range table.Indexes {
+			col := table.GetColumn(ix)
+			if col == nil {
+				continue
+			}
+			if !(col.SQLType.Name == "VARCHAR" && col.Length > maxvc) {
+				continue
+			}
+			log.Info("reducing column %s.%s from %d to %d bytes", table.Name, ix, col.Length, maxvc)
+			sqlstmt := fmt.Sprintf("alter table `%s` change column `%s` `%s` varchar(%d)", table.Name, ix, ix, maxvc)
```

Comment:
Here lies the assumption that column name == index name, which is incorrect. (Took me about 20 minutes to track this down...) Since we already have the column anyway, use `col.Name`.

But alas, the issue is deeper. `GetColumn` only returns the first column of the index, but that is not to say that there may be more. So it's actually better if instead you iterate over `table.Columns()` and then check whether it has any indexes (`len(col.Indexes) > 0`), and if it does then you proceed to reducing the col size.

Reply:
You're right, thanks. Also we not only need to check for indexes but also for primary keys as they might need to be cut down, too.